### DISCUSSION TITLE
How exactly is PFS achieved? Why can’t Meshtastic do it?

### DISCUSSION THREAD

**Discussion by trosel** [markqvist/Reticulum#531] [General]
I asked the Meshtastic community about the encryption that they have and asked if they have PFS like Reticulum. 

![IMG_1991](https://github.com/user-attachments/assets/e9c080ab-3d4c-40a9-a461-5bf159ae4bb3)

How exactly is PFS achieved in Reticulum on such a low level? And is there any way of verifying that it works today?

**Comment by faragher** [comment]
Each link session uses its own AES key. Capturing any packets from a link, then breaking AES, would not affect any other session.

I go into a little more detail here: https://github.com/markqvist/Reticulum/wiki/Frequently-Asked-Questions#how-secure-is-reticulum

If you want to verify it works, you'd need to modify Reticulum to dump the AES key when it forms a link, then send some data over the link while dumping the packets, and shut it down. Create a new link, verify the AES keys are different, then try to decrypt both dumped packets with both keys. The proof is if the packet can only be decrypted by its associated key.

As for attacking the DH key exchange directly, that's beyond my skill, you'd have to ask a crypto specialist, but it's the industry standard for good reason.

**Reply by faragher** [reply]
To expand to Meshtastic, for interoperability they use a pre-shared key, making PFS impossible unless they change keys for everyone or do a DH key exchange themselves, but having multiple people with the same key, no matter how securely shared, reduces security considerably.

**Reply by jp-bennett** [reply]
Hiya, I'm the guy under the blue redaction. Thanks for the response! How is that AES key generated, a DH operation? Are you generating a new keypair from each side of the link for each link session? Do the keys automatically rotate at some point? It's pretty interesting to see the different decisions made between the two projects. Thanks again!

Edit: Oh, and we just added a public key scheme to Meshtastic, but with essentially permanent key pairs for each node. Definitely a trade-off when it comes to things like perfect forward secrecy, but we came down on the side of saving link budget, and using the keys for identity authentication. The remote node is guaranteed to be the one you think it is. From what I understand, that's just a different design goal from Reticulum's.

**Reply by faragher** [reply]
>  From what I understand, that's just a different design goal from Reticulum's.

Absolutely. People seem to get hung up on which is "better" as opposed to them both being tools for different jobs. Hard to beat Meshtastic's ease of use and MCU-only implementation. Frankly, for most basic messaging and group chats, it's all most people would need.

Before I start linking actual implementation, broad strokes. There are two ways to run the RNS, cryptographically. First uses libraries that are pulled, IIRC, from OpenSSL, providing a rather robust and tested implementation. The second is RNS-Pure, which uses Python implementations of the cryptographic routines. It reduces dependencies on system packages, but it's also less well tested, and likely exposes more surface to attack. Reticulum itself basically doesn't care, and picks up whatever provider is available, preferring the system libraries. I think there may be some Python implementation outside of AES, but you'd have to ask Mark to be certain.

The core is based around the [handshake function](https://github.com/markqvist/Reticulum/blob/a1be97bd69e54ac10bdd90f82af1b50649b35b88/RNS/Link.py#L238) which basically performs an off-the-shelf HKDF exchange using the shared secret and link ID to generate the shared key. This is called when the link is validated, and any given link ID will alter the shared key.

That being said, I've basically exhausted my knowledge of it, and you'd have to ask Mark about the conceptual part of the encryption.

As far as automatic rotation, it depends. If the link is kept open indefinitely, the shared key will be identical. The second the link is closed and a new one established, the keys will change. At that point it's under application control and can be used based on its needs.
